File: leetcode/array/15.py
from typing import List


class Solution:

    def threeSum(self, nums: List[int]) -> List[List[int]]:
        '''a+b+c=0的不重复解
        
        @Note:
            排序,三指针,固定一个,剩余两个双边遍历
        '''
        ret = []
        lens = len(nums)
        if lens < 3:  #特殊情况
            return []
        nums = sorted(nums)  #排序
        for i in range(lens):
            a = nums[i]
            if a > 0:  #a>0,往后没解了
                break
            if i > 0 and a == nums[i - 1]:  #去除重复解
                continue
            l = i + 1
            r = lens - 1
            while l < r:
                b, c = nums[l], nums[r]
                sum = a + b + c
                if sum == 0:
                    ret.append([a, b, c])
                    while l < r and nums[l + 1] == nums[l]:  #下一个元素等于当前元素(去重)
                        l += 1
                    l += 1
                    while l < r and nums[r - 1] == nums[r]:  #下一个元素等于当前元素(去重)
                        r -= 1
                    r -= 1
                elif sum < 0:
                    l += 1
                elif sum > 0:
                    r -= 1
        return ret


if __name__ == '__main__':
    arr = [int(x) for x in input().strip().split(' ')]
    solution = Solution()
    print(solution.threeSum(arr))

File: leetcode/array/15.cpp
#include <algorithm>
#include <iostream>
#include <vector>

class Solution {
 public:
  void three_sum(std::vector<int>& arr, std::vector<std::vector<int>>* ret) {
    std::sort(arr.begin(), arr.end());

    for (size_t i = 0; i < arr.size(); ++i) {
      if (arr[i] > 0) {  // 题中跟0对比
        break;
      }
      if (i > 0 && arr[i] == arr[i - 1]) {  // 去重
        continue;
      }
      size_t l = i + 1, r = arr.size() - 1;
      while (l < r) {
        int sum = arr[i] + arr[l] + arr[r];
        if (sum == 0) {
          std::vector<int> tmp = {arr[i], arr[l], arr[r]};
          ret->emplace_back(tmp);
          while (arr[l] == tmp[1]) { // 去重
            ++l;
          }
          while (arr[r] == tmp[2]) { // 去重
            --r;
          }
        } else if (sum > 0) {
          --r;
        } else if (sum < 0) {
          ++l;
        }
      }
    }
  }
};

int main() {
  std::vector<int> a = {1, -2, 0, 2, 2, -3};

  Solution solution = Solution();
  std::vector<std::vector<int>> ret;
  solution.three_sum(a, &ret);
  for (auto& items : ret) {
    for (auto& item : items) {
      std::cout << item << " ";
    }
    std::cout << std::endl;
  }
}